add notes for nov/dec 2022 meeting (#223)
tc39/tc39-notes: meetings/2022-11/dec-01.md

MF: It doesn't make it simpler, it just stays the same. It's just that the description of Iterator.from would be simpler if you were to defer to that. So if you were to instead define flatMap flattening behavior in terms of Iterator.from, just have the exception in flatMap instead of from. I don't think it changes the complexity of describing that, just where you describe that.

YSV: On the topic of Function.prototype versus AsyncFunction.prototype. I agree with the arguments that Michael made that we should be going with function prototype and I support that. okay?

MF: That was very quiet. It sounded like support for Function.prototype for built-in async functions.

BT: Shu. agreeing with the statement. It sounds like

SYG: Yeah, that's it.

BT: Thank you, Shu.

NRO: We already have different functions like the promise combinator. So you like faction web platform the old have functions process like I would find it, strange to her different prototype, just because it deserves, like, newer methods are assumed from this.

BT: Okay. thank you. JHD. is also plus one on history,

> JHD: I just want to add that I think if async functions had been around in es6, we would have made all promise returning functions have AsyncFunction.prototype. But since we didn't, we kept with that consistency when we added a new promise combinator's and things like that. and it makes sense to keep with that consistency even though it's like I would say it's argued a little weird that built-in promise-returning functions are not AsyncFunctions, but it's not that important and we should shouldn't start doing something something different.

Reviewer: ```suggestion
JHD: I just want to add that I think if async functions had been around in es6, we would have made all promise returning functions have AsyncFunction.prototype. But since we didn't, we kept with that consistency when we added a new promise combinators and things like that. It makes sense to keep with that consistency even though I would argue it's a little weird that built-in promise-returning functions are not AsyncFunctions, but it's not that important, and we should shouldn't start doing something different.
```

JWK: That's a good suggestion. Thank you. I will try to bring this topic up in the research call.

BT: Sorry newt. there's one more item on the Queue, RPY . do you want to go ahead?

RPY: Kind of, a continuation of what SFC was saying, and so my own feeling. So if the goal here is to hide the complexity of promises then I think this is to surface level because it's really just renaming utilities that are there. I think maybe explore where users are finding that complexity and what more we could do to make it easier for people to not have to be exposed to promises. Because right now, like as soon as you debug these objects or anything you're going to be exposed to promises.

JWK: I think there isn't much space we can further hide the promises, because for things already async you don't need a promise but if things are not, for example from event emitter it has to call the promise constructor.

RPY: Okay, in that case, you may as well deliberately force users to learn promises and understand how to use them,

JWK: Hmm. Okay. Thanks.

> JHD: I have a comments about that. So for me, this proposal is not about hiding promises in any way.  I'm technically a co-champion this proposal and I think that that motivation is actually a reason not to have it. For me it's about the ergonomics of using promises, and reaching await and being able to use await.all For example, instead of having to kind of break the chain, when you do promise.all you have to do the thing with the parentheses that admittedly with something like pipeline, you would it would be a more ergonomic. My pipeline doesn't exist, yet to me this is like the same. A lot of the foot guns that I see in a way async await usage are caused by the attractiveness of using await. Unnecessarily. And when these options syntactic options are available than they become equally attractive and often they are the more correct Choice than await.

Reviewer: ```suggestion
JHD: I have a comments about that. So for me, this proposal is not about hiding promises in any way.  I'm technically a co-champion this proposal and I think that that motivation is actually a reason not to have it. For me it's about the ergonomics of using promises, and reaching for `await` and being able to use `await.all`, for example, instead of having to kind of break the chain. When you do `Promise.all` you have to do the thing with the parentheses  -that admittedly with something like pipeline, would be a more ergonomic, but pipeline doesn't exist yet. A lot of the footguns that I see in `async`/`await` usage are caused by the attractiveness of using `await` unnecessarily. And when these syntactic options are available than they become equally attractive and often they are the more correct choice than `await`.
```

JWK: await an array?

> JHD Yeah, I mean if you awake.all in a ray that is much more ergonomic and attractive than like a. So in other words what the people will do like `await a; await b; await c`, and store those in variables, what they should be doing if a, b and c depend on each other is await.all like like an array you know using a array destructuring equals await.all of an array of a promise to get out the results. Because that's you know, concurrent as and there's no there's no reason otherwise for them to be cereal. and that using awake got all is far more ergonomic and appealing like to the point where I think more users will do it rather than unnecessarily serializing their asynchronous functions. you know, like there. very sync sync functions. so, to me, that is the primary and most important aspect of this proposal. it like hiding promises. I think is very much an anti goal. Nobody should ever forget how common that promises is exist, but it should be like this to me this this avoids very, very frequent footguns.

Reviewer: ```suggestion
JHD: Yeah, I mean if you `await.all` an array, that is much more ergonomic and attractive than … in other words, when people do `await a; await b; await c`, and store those in variables, what they should be doing if `a`, `b`, and `c` depend on each other is using array destructuring equals `await.all` of an array of a promise to get out the results. Because that's concurrent,and there's no reason otherwise for them to be serial. and that using `await.all` is far more ergonomic and appealing to the point where I think more users will do it rather than unnecessarily serializing their asynchronous functions. So, to me, that is the primary and most important aspect of this proposal. Hiding promises I think is very much an anti goal. Nobody should ever forget that Promises exist, but to me this this avoids very, very frequent footguns.
```

JWK:  I can see it's possible to use DisposableStack for this use case. There is some chance that the engine can unobservably optimize. At first, we create a proxy by Proxy.revocable. Then we add the revoke function to the Disposable stack callback. The engine can drop the revoke function because it knows the revoke function is only used to revoke the proxy and if the DisposableStack is also implemented by the engine. Now revoking the proxy does not need to hold the revoker function in the memory anymore. The engine will know it should revoke the proxy when the Disposable stack calls. But yes, there are still some differences. The engine still needs to create the revoker function. It will live for a very short time but it is GC pressure.

const { revoke, proxy } = new Proxy(...)

disposableStack.defer(revoke)

return revoke

> ~~~~~~~~~~ after the return, `revoke` is no longer accessible from user land.

Reviewer 2: It breaks markdown.

Author: Should be fixed now, PTAL

# 1 December, 2022 Meeting Notes

-----

Delegates: re-use [your existing abbreviations](https://github.com/tc39/notes/blob/master/delegates.txt)! If you’re a new delegate and don’t already have an abbreviation, choose any three-letter combination that is not [already in use](https://github.com/tc39/notes/blob/master/delegates.txt), and send a PR to add it upstream.

> You can find Abbreviations in [delegates.txt](https://github.com/tc39/notes/blob/master/delegates.txt)

Reviewer 3: ```suggestion
```

YSV: Hi everyone. mmm similar to BT, I will be actually taking some time off. So we will be down one facilitator. and I think having a facilitator in the room to assist the chairs, whenever necessary is very useful. So, I'd like to encourage folks to consider becoming facilitators It is a lower bar in terms of work than doing the chair But it definitely helps the committee function smoothly and it also relieves a bit of pressure during committee time from the chairs, so that they can organize other work. that needs to be done. while the committee's in flight and make sure that things run. Let me know if you have any questions about what that looks like.

> RPR: Okay. thank you. So yeah I'll say very grateful for BT’s chairing over the years. Certainly I have learned a lot, and in any leadership situations like this there's always tough spots, there's always awkward calls and so on. So yeah it's like it's not an easy job to do. So yeah. very, grateful to BT. And also, we will miss YSV temporarily while she’s away. So yeah, so if you would enjoy the role of chair, please do reach out. We will do a formal reflector post next week, commencing the situation. Most likely we will target an election at the March meeting. So there is plenty of time because we know that something like this takes a significant number of hours. Of, and so people. need to get permission from their employers, you know, cable content to this kind of stuff is that it's ideally a long-term role. I will say that both myself and USA plan to continue in the chairing role. So, you will be supported in that time. and if yeah if you have any questions, please come to me either now and a more always happy to do so.

Reviewer 3: ```suggestion
RPR: I'll say we are very grateful for BT’s chairing over the years. Certainly I have learned a lot, and in any leadership situations like this there's always tough spots, there's always awkward calls and so on. So it's not an easy job to do. Also, we will miss YSV temporarily while she’s away. If you would enjoy the role of chair, please do reach out. We will do a formal reflector post next week, commencing the call for candidates. We will target an election at the March meeting. So there is plenty of time because we know that finding candidates takes a significant number of hours. Some people. need to get permission from their employers - this is ideally a long-term role. I will say that both myself and USA plan to continue in the chairing role. So candidates will be supported in the role. If you have any questions, please come to me either now and a more always happy to do a zoom call.
```

RBN:  and I'll say, one more thing to the discussion about, potentially deferring, the async symbol and async DisposableStack. When in the last meeting I proposed maintaining those while splitting off the rest, were primarily because I wasn't sure. We had figured out the async syntax yet. I do think that we are pretty close to that. So I'm actually more comfortable with potentially postponing those. since I hope to again, present, specifically, the async portion of this. at the next plenary. And so, I am perfectly fine with postponing those given that. We believe we have a clear path forward.

SYG: then there's nothing I need to stay with you.

> RPR: RPR: Okay. And Patrick says, plus 12 on these concerns. many to talk So next is SYG again with what's the suppressed part yet?

Reviewer 3: ```suggestion
RPR: Patrick says, plus 12 on these concerns. Many to talk So next is SYG again with "what's the suppressed part?"
```

RBN: Now, I'm sorry. I'm trying to say it's conditional on postponing async disposal stack. And symbol .s and disposed to the async using declarations part of the proposal.

SYG: So there's two conditions. There's the the deferral of the async parts to be done together with the async syntax and the reaching a resolution on JHD’s argument. Ordering is that do I understand correctly? That is correct.

WH: I also have a comment on argument ordering in `adopt`. Putting the disposer first makes it less likely that, if you get an exception while computing the disposer expression, that you fully calculate a value and never run the disposer. That’s because, with the reversed order, if computing the disposer throws, the value expression is not evaluated. It’s a factor but I’m not sure it's compelling enough to affect the order. I'm fine with this being in stage 3 with the two conditions just stated.

RPR: I just going to clarify So the the ask here is stage 3 without the AsyncDisposableStack and symbol and then we're deferring the to the argument ordering that adopted.

RBN: Yes. let's so which differing that the advancement would be conditional on that.

> RPR: Yes. it will be resolved within the state. Okay, do we have any object object, short, notice our office will ask. Do we have any support for advancing us to stage 3 based on that?

Reviewer 3: ```suggestion
RPR: Yes. it will be resolved within the state. Okay, do we have any objectiong? Do we have any support for advancing us to stage 3 based on that?
```

WH: I support it.

> RPR: We have two thumbs up in the room from PST and SFC. we have support from WH and KG on the Queue and SYG. So I got a lot of support. Do we have any any objections?

Reviewer 3: ```suggestion
RPR: We have two thumbs up in the room from PST and SFC. we have support from WH and KG on the Queue and SYG. So you have a lot of support. Do we have any any objections?
```

DE: I support advancement, can I register a request? not for not for RBN, but for I guess people who work in browsers. runs filed an issue. asking in HTML. How did you work with web platform that the proposal is long? had you know, a section in the readme describing how it would work with, with the web. I don't know if it's gotten much feedback on that part. I don't think this should block stage 3 because Ron has done this preparation and and integration proposal work. but I encourage people to get their kind of HTML departments on this too. see if they have thoughts on this. But I think that could be done later. they can this proposal makes sense whether or not various web features decided to integrate with it. So I support advancement.

RBN: Thank you.

> RPR: Then there's nothing more. Congratulations. You have stage 3 subject to the conditions. We've already said.

Reviewer 3: ```suggestion
RPR: Then there's nothing more. Congratulations. You have stage 3 subject to the conditions we've already said.
```

RBN: And I know I'm over time box. And I apologize for this but I also want to make sure that So they sync using async disposal stack. and symbol async and planning to split off. hoping to maintain stage 2 for that. And I'm also plenty of put off void bindings. I'm that I'm I'm fine with it. Actually re-proposing. It's stage one. one. But I want to make sure that We don't. have any. objections to maintaining stage 2 for those features that were splitting off.

DE: Sorry, I'm not really sold on void bindings being stage 2. I think we probably should go with another approach. That's not based on an operator, but using it in a not operator kind of syntax. but I think it's important, it's a really important problem to solve.

RBN: So, Mark that. one out. Like I said, I'm happy to re-propose that at stage one if necessary. So but at the very least I think there was General preference for pushing this. Pushing the async using, I just want to make sure that if I can to maintain stage 2 for that. Side of the proposal. Are there any objections?

KG: I explicitly support keeping it in stage 2.

RBN: Thank you.

> RPR: Alright. I think that part is kept at stage 2.

Reviewer 3: ```suggestion
RPR: Alright. That part is kept at stage 2.
```

RBN: I'm not sure if Daniel might have any other concerns, but I might say that I won't really block stage 2, but I do think this is a concern, to, to consider. I know that the earlier bleep was dominant in a cola is Luca’'s proposal which was somewhat similar as providing a block mechanism. That was portable did close over things, but you would have to supply those close value. So that might be something we need to talk more. or as we're pushing this up for stage between stage, 2, and stage 3.

KKL:  and, for the record, I also will not block but I do think that in stage 3 and Stage 2, we will run into physical limitations that will not allow for the possibility of preserving the invariant the constructing, a new module from its source always works. but

DE: I think that will be fine. And I think we can discuss that later.

> RPR: There is any Objections. to this, to block stage 2 ? . All right. Is there any Okay ? Do we have any support for stage 2?

Reviewer 3: ```suggestion
RPR: Are there any objections that would block stage 2? Do we have any support for stage 2?
```

GB:, it very well thought out in line with the other modules work that's been going on. I do have stage three concerns but I don't know any stage to consume. I suppose. for stage 2.

> RPR:   USA in the room with plus 1.

Reviewer 3: ```suggestion
RPR:  USA in the room with +1
```

> RPR: I can see that there are concerns on the queue Daniel did you want to speak to yours? Which want to make sure that we before everything out before

Reviewer 3: ```suggestion
RPR: I can see that there are concerns on the queue. Daniel did you want to speak to yours? We want to make sure that we bring everything out before concluding.
```

DRR: I think just the thing about having a world of values and then module values that you need To sort of treat separately. doesn't quite feel right? to me. It's something that I've also voice in terms of. earlier forms of decorators when they look more like macros but so I don't think that's a stage 2 blocker but it may be for moving forward past that.

RPR: Okay, thank you.

JWK: This is different than the decorators case in the static decorator, if you Imports it's you cannot use it in the normal random value space.

DRR:  but this although can cure but you can't, you have a special form of values that you can import from, but you can't do that with any other value. Yeah. so your yeah. I mean you're creating a new thing that you can do with Imports but only certain things can be done. You can do that with. so that's where kind of fun I don't want to spend too much time on this,

> RPR: All right, thank you. Waldemar. Can you go in like 20 seconds?

Reviewer 3: ```suggestion
RPR: All right, thank you. WH. Can you go in 20 seconds?
```

WH: Yes; Maybe I just don't understand this, but I don't know what happens if you have a module declaration inside a function. When does it take effect and become visible to other modules? What if you have one inside a loop? What other things can refer to it? So I just don't yet understand the interaction of module scoping with our language scoping here.

### Conclusion/Resolution

State 2

Lots of open topics

> RPR: It's something that needs to be clearly spelled out in the spring. All right. Okay so we have to have that. So in conclusion, we have stage  2  for this  with a lot of  remaining topics. right. Thank you. okay, moving on. So we are a bit behind time now. because we need to the next topic is 60 minutes. So therefore this meeting will finish. at 4:17 if if this topic does need its full time but that is the final topic of the day. so, caridy, are you here to present? ShadowRealms?  So before we proceed, please, you need another note take for the final session. Thank you.

Reviewer 3: ```suggestion
```

SYG: You said web reality. Is this stuff shipped?

SFC: Yeah, All the browser's are shipping this now. so this is about web reality.

SYG: All right. It seems like the safest path forward now is to require the minimum I guess.

USA: I also wanted to say that given that we already have feedback from implementors, we don't have any reason to wait until stage 4.

SYG Is there guidance on the minimum? is the minimum meant to be aligned with ICU4X changed?

SFC:  that, that was the intent except that now that Chrome gives us what a minimum might be. We might just match what that is.

SYG: I mean, I think there's some wiggle room depending on how recently we shipped it. What we think that usage numbers are like I don't think this is as set in stone but I'll defer to folks doing the actual field research in the FYT here.

EAO : +1 for the 1st option

SFC: Great. So if we're all okay, with having a minimum limit that will move forward and you know, make that change to the specification.

> RPR:  another plus one from DE.

Reviewer 3: ```suggestion
RPR:  Another +1 from DE.
```

> RPR: Excellent. Thank you. Everyone. Thank you, SFC. All right. And with that we offer fully complete …
> 
> RPR lists the proposal deferred to the next meeting:
> 
> 
> 

Reviewer 3: ```suggestion
RPR lists the proposal deferred to the next meeting:

```
